accepted/00063-hub-iss.md

```diff
@@ -123,7 +136,7 @@ At this point it is believed/assumed that:
 
```

Comment:
At line 134, that statement is not true any more. The linking tables has foreign keys to other tables, so we can replace the ID'd using the same mechanism used in the foreign keys. This approach is implemented and tested in the POC.

Reply:
Just out of excessive care: was it tested in the case one channel used to have, say, 3 packages and now only has 2? Will it be reflected correctly after a new sync round?

Reply:
If the channel has fewer packages it will not be covered with this implementation. In this case, we need to think about how to deal it.

Reply:
I did reproduce this case with the original ISS and indeed the extra package is removed. I think we should make sure this actually happens the same with the new implementation proposed here.

The idea I had is to issue a `DELETE` statement removing any row which is now unnecesary. I see two alternatives:
1. mass-delete all rows in a table that are affected by this ISS round (eg. `DELETE FROM rhnChannelPackage WHERE channel_label = 'label'`), before any `INSERT` (`ON CONFLICT` clause could in principle be omitted in this case)
2. selectively delete rows with IDs not in the set we are looking at (eg. `DELETE FROM rhnChannelPackage WHERE channel_id, package_id IN (...possibly complicated subselect here...)). Probably slower/more difficult to implement

Some RDBMSs support another (standard) SQL command called `MERGE INTO`, which seems to be doing what we want (INSERT, UPDATE or DELETE depending on conditions). This was scheduled to be merged into PG 11, but [it got reverted](08ea7a2291db21a618d19d612c8060cda68f1892) and indeed we still do not have it in PG 13. Even looking at the [roadmap](https://www.postgresql.org/developer/roadmap/), I see no traces left of it, so I would not consider it as a possibility for the foreseeable future.

At this point I am out of other viable ideas (but would love to hear more, if anybody has).

Reply:
We should also consider that the delete will depend on the data already on the target database, which can be different between the target servers.
Link tables can be deleted with delete all, and re-inserted. Should we also remove entries from `rhnpackage` or `rhnerrata`, or it's enough to remove from the linking tables, like `rhnchannelpackage` and `rhnerratapackage`?

Reply:
It depends on the nature of the table's data. I tried to compile a list looking at what the current ISS implementation does and some common sense from my part (hopefully right): tables that need cleanup are marked `// cleanup` here:

https://github.com/uyuni-project/inter-server-sync/blob/360a110b4578fa29dec62788e85f5c3c82f9ca8d/dumper/channelDataDumper.go#L29-L86

As you can see, tables like `rhnErrata` and `rhnPackage` can be left alone and most of linking tables need cleanup, but there is also some others that do not necessarily fit each definition. Ultimately, cleaning never hurts - it's just not necessarily useful.

At the end of the day, implementing a general mechanism will be needed, and we can later judge if we should just mass-apply it to all tables or to a selected inclusion / exclusion list.

From an RFC POV I think the text in line 134 can stay, or do you see a need for any other rewording?

Reply:
I agree, we need some kind of cleanup mechanism for those linking tables